lib/algos/list/delimiterMaster.ts

```diff
@@ -177,14 +437,124 @@ export class DelimiterMaster extends Delimiter {
         return this.filter_onNewMasterKeyV0(key, value);
     }
 
+    keyHandler_SkippingGapV0(key: string, value: string): FilterReturnValue {
+        const { gapCache, gapCached } = <GapCachingInfo_GapCached> this._gapCaching;
+        if (key <= gapCached.lastKey) {
+            // Temporarily create a gap with weight=0 from the latest
+            // skipped key. This will allow to extend the existing gap
+            // being skipped with a chained gap in case other delete
+            // markers are seen after the existing gap is skipped.
+            this._createBuildingGap(key, 0);
+            return FILTER_SKIP;
+        }
+        this._gapCaching = {
+            state: GapCachingState.UnknownGap,
+            gapCache,
+        };
+        this.setState(<DelimiterMasterFilterState_SkippingVersionsV0> {
+            id: DelimiterMasterFilterStateId.SkippingVersionsV0,
+        });
+        return this.handleKey(key, value);
+    }
+
     skippingBase(): string | undefined {
         switch (this.state.id) {
         case DelimiterMasterFilterStateId.SkippingVersionsV0:
             const { masterKey } = <DelimiterMasterFilterState_SkippingVersionsV0> this.state;
             return masterKey + inc(VID_SEP);
 
+        case DelimiterMasterFilterStateId.SkippingGapV0:
+            const { gapCached } = <GapCachingInfo_GapCached> this._gapCaching;
+            return gapCached.lastKey;
+
         default:
             return super.skippingBase();
         }
     }
+
+    result(): ResultObject {
+        this._cutBuildingGap();
+        return super.result();
+    }
+
+    _saveBuildingGap(): void {
+        const { gapCache, params, gap, gapWeight } =
+              <GapBuildingInfo_Building> this._gapBuilding;
+        const totalElapsed = Date.now() - params.initTimestamp;
+        if (totalElapsed >= gapCache.exposureDelayMs) {
+            this._gapBuilding = {
+                state: GapBuildingState.Disabled,
+            };
+            return;
+        }
+        const { firstKey, lastKey, weight } = gap;
+        gapCache.setGap(firstKey, lastKey, weight);
+        this._gapBuilding = {
+            state: GapBuildingState.Building,
+            gapCache,
+            params,
+            gap: {
+                firstKey: gap.lastKey,
+                lastKey: gap.lastKey,
+                weight: 0,
+            },
+            gapWeight,
+        };
```

Comment:
Yes it is on purpose, the `initTimestamp` can only be set when the listing starts, we're not allowed to change it until a new listing request is made. This is because the listing uses a snapshot on the DB so it returns the entries in the state they were in at that time.

This also means that we will need to restart the listing periodically to overcome this, that's a reason to have an `expirationDelayMs` not too small (thinking maybe 1 second to 10 seconds would be appropriate). Triggering listing requests is normally inexpensive, what gets expensive is listing thousands of entries. But the code in repd will need to be adapted for this, and call `refreshGapCache()` periodically every time a new listing is about to be triggered.

Reply:
How does this work? How can the listing work on a snapshot of the DB? Does it rollback changes that happened during the listing? Is it related to the `WriteCache` you mentioned when we discussed this design? 

I'd like to understand this better so that I can take this into account when we will tackle metadata sharding.

Reply:
It's all handled by the LevelDB engine. LevelDB has also explicit snapshot support with an API, but for listings it automatically uses an internal snapshot of the DB.

Reply:
That's pretty nifty! Makes sense that the LSM-structure easily allows such snapshot feature, is a snapshot created every time that a listing starts? Do we really need the listing to be done on a snapshot though? Sounds like it would be borderline impossible to do on a sharded metadata model, I was under the impression that the listing had to contain entries that represent all updates until the beginning of the listing but not necessary represent an instant in time. What do you think? 

Maybe we can keep this conversation going in Slack or somewhere more appropriate, I want to present to you an update on the designs we discussed for metadata sharding :)